core/src/main/scala/kafka/server/metadata/MetadataPartitions.scala

```diff
@@ -39,11 +41,30 @@ object MetadataPartition {
       record.isr(),
       Collections.emptyList(), // TODO KAFKA-12285 handle offline replicas
       Collections.emptyList(),
-      Collections.emptyList())
+      Collections.emptyList(),
+      largestDeferredOffsetEverSeen = deferredAtOffset.getOrElse(OffsetNeverDeferred),
+      isCurrentlyDeferringChanges = deferredAtOffset.isDefined)
```

Comment:
Hmm... why do we need this boolean?  Can't we just check if `largestDeferredOffsetEverSeen` is not `OffsetNeverDeferred`

Reply:
We basically use `largestDeferredOffsetEverSeen` only for logging at this point -- we also check it in a few `private def sanityCheckState...()` `RaftReplicaManager` methods.  We could completely eliminate `largestDeferredOffsetEverSeen` if we didn't want to log when the partition was last deferred.  It just tracks when the partition was last seen and the change at that offset was deferred rather than directly applied.  Once the partition is no longer deferred the value remains whatever it was and the boolean flips to `false`.

It does seem on the surface that we could change the declaration to `deferredSinceOffset` and get rid of the boolean -- and `deferredSinceOffset` would change to `-1` once those changes are applied.  But there is a problem with this if the partition changes to not being deferred in the metadata cache before we ask `RaftReplicaManager` to process all of its deferred changes: the value will be -1 in the metadata cache under those circumstances, and we wouldn't have the value to log.

So I think we have a few options.

1. Do the logging, apply the changes to the matadata cache before replica manager, and keep the `Long` and `Boolean` as currently defined
2. Do the logging, apply the changes to the matadata cache **after** replica manager, and use just a `Long` (with the semantics being changed as described above)
3. Just use a Boolean and don't do the logging.

Reply:
As discussed offline, we will eliminate the information from the messages we log when applying deferred changes, and we won't carry that info around in `MetadataPartition`.  Currently `RaftReplicaManager` knows if it is deferring changes or not.  Maybe later when we get `BrokerLifecycleManager` and `BrokerMetadataListener` committed we can think about where a global boolean might live to identify if the broker is fenced or not.  It isn't critical to decide right now because we are only going to defer the application of partition metadata at startup in 2.8.